### PR TITLE
Fixes a control problem occurring in rare circumstances for the fixed fan speed, variable water flow fancoil

### PR DESCRIPTION
**Pull request by wfbuhl**
## Pull request overview

Addresses #5276. In cases where the plant locks flow on the demand side, an internal component bypass allows the fan coil to still vary it's output to attempt to meet the load (within available water flow rate).
### Work Checklist

Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
- [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
- [ ] At least one of the following appropriate labels must be added to this PR to be consumed into the changelog:
  - Defect: This pull request repairs a github defect issue.  The github issue should be referenced in the PR description
  - Refactoring: This pull request includes code changes that don't change the functionality of the program, just perform refactoring
  - NewFeature: This pull request includes code to add a new feature to EnergyPlus
  - Performance: This pull request includes code changes that are directed at improving the runtime performance of EnergyPlus
  - DoNoPublish: This pull request includes changes that shouldn't be included in the changelog
### Review Checklist

This will not be exhaustively relevant to every PR.
- [x] Code style (parentheses padding, variable names)
- [x] Functional code review (it has to work!)
- [x] If defect, results of running current develop vs this branch should exhibit the fix
- [x] CI status: all green or justified
- [ ] Performance: CI Linux results include performance check -- verify this
- [x] Unit Test(s)
- C++ checks:
  - [ ] Argument types
  - [ ] If any virtual classes, ensure virtual destructor included, other things
- IDD changes:
  - [ ] Verify naming conventions and styles, memos and notes and defaults
  - [ ] If transition, add rules to spreadsheet
  - [ ] If transition, add transition source
  - [ ] If transition, update idfs
- [ ] If new idf included, locally check the err file and other outputs
- [ ] Documentation changes in place
- [ ] ExpandObjects changes??
- [ ] If output changes, including tabular output structure, add to output rules file for interfaces

### CODE REVIEW

File: src/EnergyPlus/FanCoilUnits.cc [hunk 1]
```diff
@@ -1845,8 +1845,8 @@ namespace FanCoilUnits {
 		using General::TrimSigDigits;
 		using PlantUtilities::SetComponentFlowRate;
 		using General::SolveRegulaFalsi;
-		using DataZoneEquipment::CalcDesignSpecificationOutdoorAir;
-
+		using DataZoneEquipment::CalcDesignSpecificationOutdoorAir;		using namespace DataPlant;
+		using namespace DataLoopNode;
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc:1850]: wrap issue

File: src/EnergyPlus/FanCoilUnits.cc [hunk 2]
```diff
@@ -1936,61 +1950,152 @@ namespace FanCoilUnits {
...
+					SolveRegulaFalsi( 0.001, MaxIterCycl, SolFlag, CWFlow, CalcFanCoilCWLoadResidual, 0.0, MaxWaterFlow, Par );
+					if ( SolFlag == -1 ) {
+						ShowWarningError( "Cold Water control failed in fan coil unit " + FanCoil( FanCoilNum ).Name );
+						ShowContinueError( "  Iteration limit exceeded in calculating water flow rate " );
+						ShowRecurringWarningErrorAtEnd( "Cold water flow Iteration limit exceeded in fan coil unit " + FanCoil( FanCoilNum ).Name, FanCoil( FanCoilNum ).MaxIterIndexH );
+					}
+					else if ( SolFlag == -2 ) {
+						ShowWarningError( "Cold Water control failed in fan coil unit " + FanCoil( FanCoilNum ).Name );
+						ShowContinueError( "  Bad cold water mass flow limits" );
+						ShowRecurringWarningErrorAtEnd( "Cold Water control failed in fan coil unit " + FanCoil( FanCoilNum ).Name, FanCoil( FanCoilNum ).MaxIterIndexC );
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc]: There is no limit on the warning and continue errors and could lead to very large err files.

File: src/EnergyPlus/FanCoilUnits.cc [hunk 3]
```diff
@@ -1936,61 +1950,152 @@ namespace FanCoilUnits {
...
+				}
+				if ( ! ColdFlowLocked ) {
+					mdot = CWFlow; // not flowlocked - set flow to CWFlow
+					SetComponentFlowRate( mdot, FanCoil( FanCoilNum ).ColdControlNode, FanCoil( FanCoilNum ).ColdPlantOutletNode, FanCoil( FanCoilNum ).CWLoopNum,
+						FanCoil( FanCoilNum ).CWLoopSide, FanCoil( FanCoilNum ).CWBranchNum, FanCoil( FanCoilNum ).CWCompNum );
+					Calc4PipeFanCoil( FanCoilNum, ControlledZoneNum, FirstHVACIteration, QUnitOut ); // get QUnitOut
+				}
+				else {
+					// flow lock on 
+					if ( MdotLockC > CWFlow ) { // if mdot > CWFlow, bypass extra flow
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc:2038]: I'm not familiar with a component bypass, where is the excess water being sent?

File: src/EnergyPlus/FanCoilUnits.cc [hunk 4]
```diff
@@ -1936,61 +1950,152 @@ namespace FanCoilUnits {
...
+							ShowWarningError( "Hot Water control failed in fan coil unit " + FanCoil( FanCoilNum ).Name );
+							ShowContinueError( "  Iteration limit exceeded in calculating water flow rate " );
+							ShowRecurringWarningErrorAtEnd( "Hot water flow Iteration limit exceeded in fan coil unit " + FanCoil( FanCoilNum ).Name, FanCoil( FanCoilNum ).MaxIterIndexH );
 						}
-						ControlCompOutput( FanCoil( FanCoilNum ).Name, cMO_FanCoil, FanCoilNum, FirstHVACIteration, QZnReq, ControlNode, MaxWaterFlow, MinWaterFlow, ControlOffset, FanCoil( FanCoilNum ).ControlCompTypeNum, FanCoil( FanCoilNum ).CompErrIndex, _, _, _, iNormalAction, _, FanCoil( FanCoilNum ).HWLoopNum, FanCoil( FanCoilNum ).HWLoopSide, FanCoil( FanCoilNum ).HWBranchNum, ControlledZoneNum );
-					} else {
+						else if ( SolFlag == -2 ) {
+							ShowWarningError( "Hot Water control failed in fan coil unit " + FanCoil( FanCoilNum ).Name );
+							ShowContinueError( "  Bad hot water mass flow limits" );
+							ShowRecurringWarningErrorAtEnd( "Hot Water control failed in fan coil unit " + FanCoil( FanCoilNum ).Name, FanCoil( FanCoilNum ).MaxIterIndexH );
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc]: Same issue on excess messages to err file

File: src/EnergyPlus/FanCoilUnits.cc [hunk 5]
```diff
@@ -1998,6 +2103,40 @@ namespace FanCoilUnits {
 						Par( 4 ) = QZnReq;
 						SolveRegulaFalsi( 0.001, MaxIterCycl, SolFlag, PLR, CalcFanCoilLoadResidual, 0.0, 1.0, Par );
 					}
+				} else {
+					// demand greater than capacity
+					if ( FanCoil( FanCoilNum ).HCoilType_Num == HCoil_Water ) {
+						HWFlow = MaxWaterFlow;
+					} else{
+						Calc4PipeFanCoil( FanCoilNum, ControlledZoneNum, FirstHVACIteration, QUnitOut, 1.0 );
+					}
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc:2133]: What does this else do? If demand > capacity, yes to full flow. But don't you need to call Calc4PipeFanCoil regardless?

File: tst/EnergyPlus/unit/FanCoilUnits.unit.cc [hunk 6]
```diff
@@ -1114,5 +2125,4 @@ namespace EnergyPlus {
 		CoilNames.deallocate();
 
 	}
-
-}
+}
```
Comment (rraustad) [inline tst/EnergyPlus/unit/FanCoilUnits.unit.cc:2122]: Missing final carriage return.

This all looks sound!

File: src/EnergyPlus/FanCoilUnits.cc [hunk 7]
```diff
@@ -1770,7 +1770,7 @@ namespace FanCoilUnits {
 					FanCoil( FanCoilNum ).DesZoneCoolingLoad = -1.0 * FanCoil( FanCoilNum ).DesCoolingLoad;
 					FanCoil( FanCoilNum ).DesZoneHeatingLoad = FanCoil( FanCoilNum ).DesHeatingLoad;
 
-				}
+		}
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc:1773]: Justification?

File: src/EnergyPlus/FanCoilUnits.cc [hunk 8]
```diff
@@ -1936,61 +1950,152 @@ namespace FanCoilUnits {
...
 			}
+			// obtain unit output with no active heating/cooling
 			Calc4PipeFanCoil( FanCoilNum, ControlledZoneNum, FirstHVACIteration, QUnitOutNoHC, 0.0 );
-
+			if ( FirstHVACIteration ) {
+				FanCoil( FanCoilNum ).QUnitOutNoHC = QUnitOutNoHC;
+			}
+			else {
+				QUnitOutNoHC = FanCoil( FanCoilNum ).QUnitOutNoHC;
+			}
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc:1969]: If your going to do this, why not wrap all the way up to 1951 and not do those for !FirstHVACIteration.

```
if ( FirstHVACIteration ) {
  line 1951 down to 1964 inserted 
  FanCoil( FanCoilNum ).QUnitOutNoHC = QUnitOutNoHC;
 } else {
  QUnitOutNoHC = FanCoil( FanCoilNum ).QUnitOutNoHC;
 }
```

File: src/EnergyPlus/FanCoilUnits.cc [hunk 9]
```diff
@@ -1936,61 +1950,152 @@ namespace FanCoilUnits {
...
 				SetComponentFlowRate( mdot, FanCoil( FanCoilNum ).ColdControlNode, FanCoil( FanCoilNum ).ColdPlantOutletNode, FanCoil( FanCoilNum ).CWLoopNum, FanCoil( FanCoilNum ).CWLoopSide, FanCoil( FanCoilNum ).CWBranchNum, FanCoil( FanCoilNum ).CWCompNum );
-				Calc4PipeFanCoil( FanCoilNum, ControlledZoneNum, FirstHVACIteration, QUnitOutMaxHC );
-				if ( QUnitOutMaxHC < QCoilCoolSP ) {
+				Calc4PipeFanCoil( FanCoilNum, ControlledZoneNum, FirstHVACIteration, QUnitOutMaxC );
+				if ( ! ColdFlowLocked ) {
+					FanCoil( FanCoilNum ).QUnitOutMaxC = QUnitOutMaxC;
+				} else {
+					QUnitOutMaxC = FanCoil( FanCoilNum ).QUnitOutMaxC;
+					MdotLockC = mdot; // save locked flow
+				}
```
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc:1994]: Same comment as line 1970

File: src/EnergyPlus/FanCoilUnits.cc (hunk 5 above)
Comment (rraustad) [inline src/EnergyPlus/FanCoilUnits.cc:2133]: OK, it looks like you have already called Calc4PipeFanCoil at full flow at line 2062

File: src/EnergyPlus/FanCoilUnits.cc (hunk 2 above)
Comment (wfbuhl) [inline src/EnergyPlus/FanCoilUnits.cc]: Same for your code I think! I'll add an error counter and a check. Also a timestamp.

File: src/EnergyPlus/FanCoilUnits.cc (hunk 9 above)
Comment (wfbuhl) [inline src/EnergyPlus/FanCoilUnits.cc:1994]: Guess I don't understand your comments

File: src/EnergyPlus/FanCoilUnits.cc (hunk 8 above)
Comment (wfbuhl) [inline src/EnergyPlus/FanCoilUnits.cc:1969]: Need to check for flow lock each time through and use saved value of no flow output when plant has locked the flow. When the flow is locked SetComponentFlowRate gives back the locked value, and doesn't set the requested value (of zero). Thus I don't get the no flow output, but something else. For FirstHVACIteration I know I am always getting the correct QUnitOutNoHC.

File: src/EnergyPlus/FanCoilUnits.cc (hunk 3 above)
Comment (wfbuhl) [inline src/EnergyPlus/FanCoilUnits.cc:2038]: This is the kludge that fixes the bug. In the defect file (unreproducible elsewhere) plant flows are getting locked down due to flow imbalance in plant between supply and demand. This is a large flow imbalance and I could find no reason for it. With the flow locked, the constant fan variable flow fan coil can't respond to the load. So I stuck in an internal bypass - the flow through the coil can still vary but the total flow is constant at the locked value. The bypassed water rejoins at the outlet node.

File: src/EnergyPlus/FanCoilUnits.cc (hunk 4 above)
Comment (wfbuhl) [inline src/EnergyPlus/FanCoilUnits.cc]: ditto

**Comment by rraustad** [comment]
These results don't make sense. There should be no diff's in non-fan coil files. I switched to this branch and pulled. It said the branch was up to date.

**Comment by mjwitte** [comment]
It's behind develop by 51 commits.  Needs a merge.

**Comment by mjwitte** [comment]
Look here https://github.com/NREL/EnergyPlus/branches/active

**Comment by rraustad** [comment]
Diff's (eplusout.csv.absdiff) for DOAToFanCoilInlet shows a 1.4C difference in zone 2 (and 4) temp for the winter design day - all day. Comparing V8.4 to this branch shows only a small difference which does not occur the entire day. The FC looks to be operating similarly to V8.4.

issue5276_doatofancoilinlet

**Comment by rraustad** [comment]
Another large diff file that looks similar to before.

issue5276_fancoilautosizescalablesizing

**Comment by rraustad** [comment]
Waiting for CI and will accept tomorrow if clean.

**Comment by rraustad** [comment]
Where is the defect file for #5276?

**Comment by rraustad** [comment]
Fred, I question the mixing equation. Why does the enthalpy of the mixed water stream remain the same as that leaving the fan coil at the coil water outlet node? For water, if T changes, so does H. If H at the mixed stream is the same as the fan coil water outlet, then when there is bypass the total Q is greater than Qfancoil. What am I missing?

```
// note outlet temperature changes but outlet enthalpy stays the same
Node( FanCoil( FanCoilNum ).ColdPlantOutletNode ).Temp = ( CWFlowBypass * Node( FanCoil( FanCoilNum ).ColdControlNode ).Temp +
                        CWFlow * Node( FanCoil( FanCoilNum ).ColdPlantOutletNode ).Temp ) / MdotLockC;
```

**Comment by rraustad** [comment]
Checked HAMT file based on comment above where results showed odd diffs.

hamt_hourly_5276

**Comment by rraustad** [comment]
Other "odd diff's" file tested with same results as V8.4. 
Nearly ready for check in. Waiting only on defect file and answer to water mixing equations.

hvactemplate_fancoil_5276

**Comment by wfbuhl** [comment]
I deliberately didn't change the outlet enthalpy. Now why was that? Doesn't
make sense now. I'll let it fester for a few hours. Defect file - I had one
(shortened version of original). Thought I uploaded it. I'll check on that.

Fred

**Comment by rraustad** [comment]
Defect file confirmed that fix correct the problem. Before and after shown here.

5276-defectfile

**Comment by rraustad** [comment]
This issue is complete except for the mixing equation for enthalpy. Same as for T, unless there is a water enthalpy function available.

**Comment by wfbuhl** [comment]
I uploaded the defect file I was using and the weather file. Since Brent
uploaded the original defect file and you have run it, you can ignore these.

I believe I was confused previously when I didn't alter the enthalpy. I'll
go ahead and do that now and recommit.

Fred

**Comment by rraustad** [comment]
Well, I can't explain this since I believe there is bypass flow at this point in the simulation. The water outlet enthalpy appears to agree with the fan coil air-side load before and after Fred's water outlet enthalpy correction. Regardless of this, I am comfortable with these changes.

issue5276-waterenthalpy

**Comment by rraustad** [comment]
I'll wait until the CI finishes to merge this in.

**Comment by rraustad** [comment]
@Myoldmopar are these CI machines real or virtual? Are these run on blades or actual desktops?

**Comment by mbadams5** [comment]
@rraustad What do you mean "are these CI machines real or virtual"? The CI machines are a mix of servers and desktops. In my case, I have 3 Windows desktop machines running and 1 Ubuntu server. I know Kyle's Mac is a desktop machine. I believe the CI machines at NREL are all desktops, but some of them are virtual machines on the same physical machine. Does that answer your question? And why are you asking?

**Comment by rraustad** [comment]
Just wanted to know. I have access to blades where these could all be run in a virtual environment.

**Comment by mbadams5** [comment]
I'll let you and @Myoldmopar talk about that.

**Comment by mjwitte** [comment]
@rraustad Please edit the description at the top and check off the review items.

**Pull request merged**